coderd/database/dbmem/dbmem.go

```diff
@@ -1180,14 +1183,72 @@ func (q *FakeQuerier) DeleteOAuth2ProviderAppByID(_ context.Context, id uuid.UUI
 			q.oauth2ProviderApps[index] = q.oauth2ProviderApps[len(q.oauth2ProviderApps)-1]
 			q.oauth2ProviderApps = q.oauth2ProviderApps[:len(q.oauth2ProviderApps)-1]
 
-			secrets := []database.OAuth2ProviderAppSecret{}
+			// Cascade delete into secrets.
+			var deletedSecretIDs []uuid.UUID
+			var secrets []database.OAuth2ProviderAppSecret
 			for _, secret := range q.oauth2ProviderAppSecrets {
-				if secret.AppID != id {
+				if secret.AppID == id {
+					deletedSecretIDs = append(deletedSecretIDs, secret.ID)
+				} else {
 					secrets = append(secrets, secret)
 				}
 			}
 			q.oauth2ProviderAppSecrets = secrets
 
+			// Cascade delete into tokens.
+			var keyIDsToDelete []string
+			var tokens []database.OAuth2ProviderAppToken
+			for _, token := range q.oauth2ProviderAppTokens {
+				if slice.Contains(deletedSecretIDs, token.AppSecretID) {
+					keyIDsToDelete = append(keyIDsToDelete, token.APIKeyID)
+				} else {
+					tokens = append(tokens, token)
+				}
+			}
+			q.oauth2ProviderAppTokens = tokens
+
+			// Delete from API keys.
+			var keys []database.APIKey
+			for _, key := range q.apiKeys {
+				if !slices.Contains(keyIDsToDelete, key.ID) {
+					keys = append(keys, key)
+				}
+			}
+			q.apiKeys = keys
+
+			return nil
+		}
+	}
+	return sql.ErrNoRows
+}
+
+func (q *FakeQuerier) DeleteOAuth2ProviderAppCodeByID(_ context.Context, id uuid.UUID) error {
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	for index, code := range q.oauth2ProviderAppCodes {
+		if code.ID == id {
+			q.oauth2ProviderAppCodes[index] = q.oauth2ProviderAppCodes[len(q.oauth2ProviderAppCodes)-1]
+			q.oauth2ProviderAppCodes = q.oauth2ProviderAppCodes[:len(q.oauth2ProviderAppCodes)-1]
+			return nil
+		}
+	}
+	return sql.ErrNoRows
+}
+
+func (q *FakeQuerier) DeleteOAuth2ProviderAppCodesByAppAndUserID(_ context.Context, arg database.DeleteOAuth2ProviderAppCodesByAppAndUserIDParams) error {
+	err := validateDatabaseType(arg)
+	if err != nil {
+		return err
+	}
+
+	q.mutex.Lock()
+	defer q.mutex.Unlock()
+
+	for index, code := range q.oauth2ProviderAppCodes {
+		if code.AppID == arg.AppID && code.UserID == arg.UserID {
+			q.oauth2ProviderAppCodes[index] = q.oauth2ProviderAppCodes[len(q.oauth2ProviderAppCodes)-1]
+			q.oauth2ProviderAppCodes = q.oauth2ProviderAppCodes[:len(q.oauth2ProviderAppCodes)-1]
```

Comment:
Good point!  I think we will probably not have to worry about it since we do not rely on ordering of authentication codes (there is no method to get multiple codes right now either) but seems worth keeping in mind.

I lazily copied these lines from other implementations in this file but I think it might be better to add a `remove` and change all of these:

```
func remove[T any](slice []T, i int) []T {
    return append(slice[:i], slice[i+1:]...)
}
```

Reply:
The way I have been doing it (for apps and secrets which do need ordering) is to sort them before returning in their `Get*` calls so the underlying slice can be in any order.

Reply:
Ooo 1.21 has a `Delete` for slices apparently.